packages/server/src/db/tenant_creation.js

```diff
@@ -0,0 +1,235 @@
+    if (missingKeys.length > 0) {
+        return `Missing required options: ${missingKeys.join(', ')}`;
+    }
+    if (extraKeys.length > 0) {
+        return `Unknown options: ${extraKeys.join(', ')}`;
+    }
+
+    const errors = (await Promise.all(questions.map(async (question) => {
+        const validator = question.validate || (() => true);
+        const valid = await validator(options[question.name]);
+        return valid !== true ? valid : null;
+    }))).filter(_.identity);
+
+    if (errors.length > 0) {
+        return errors.join('; ');
+    }
+
+    return true;
+}
+
+async function promptForCreateTenantOptions(trns = knex) {
+    const questions = buildInquirerQuestionsForCreateTenantOptions(trns);
+    const { confirmed, ...options } = await inquirer.prompt([
+        ...questions,
+        {
+            name: 'confirmed',
+            type: 'confirm',
+            message: (answers) => {
+                console.log(answers);
+                return 'Everything look good?';
+            },
+        },
+    ]);
+
+    if (!confirmed) {
+        console.log('Aborting');
+        process.exit(0);
+        return null;
+    }
+
+    return options;
+}
+
+async function createTenant(options, trns = knex) {
+    const optionsValid = await validateCreateTenantOptions(options, trns);
+    if (optionsValid !== true) {
+        throw new Error(`invalid create tenant options: ${optionsValid}`);
+    }
+    const {
+        tenantName, agencyName, agencyAbbreviation, agencyCode, adminUserEmail, adminUserName,
+    } = options;
+
+    // Seeded tenants, agencies, and users with fixed IDs can screw up the autoincrement, apparently
+    await Promise.all(['tenants', 'agencies', 'users'].map((tableName) => trns.raw(`select setval('${tableName}_id_seq', max(id)) from ${tableName}`)));
+
+    // Create tenant
+    const { tenantId } = await trns('tenants')
+        .insert({
+            display_name: tenantName,
+            // main_agency_id null because we haven't yet created the agency, so don't know its ID
+            main_agency_id: null,
+        })
+        .returning('id as tenantId')
+        .then((rows) => rows[0]);
+
+    // Create root agency
+    // We don't use db.createAgency because it expects a creatorId, but we are
+    // creating in a new tenant that has no users yet.
+    const { agencyId } = await trns('agencies')
+        .insert({
+            name: agencyName,
+            abbreviation: agencyAbbreviation,
+            code: agencyCode,
+            parent: null,
+            // main_agency_id is non-nullable, but trns.ref('id') doesn't seem to work. So we have
+            // to set to an existing value, then update it afterward.
+            //
+            // TODO(mbroussard): should we remove the NOT NULL constraint and have null mean root
+            // agency vs. having self-loops (similar to parent field)?
+            main_agency_id: await firstAgencyId(trns),
```

Comment:
This is definitely awkward.  And it should be redundant with the `main_agency_id` field in the `tenants` table, is that correct?

I would definitely prefer to land a migration that removes this redundancy and updates any queries rather than find a weird workaround for this self-reference.

Reply:
I think you're right that these fields are redundant and we could get rid of the one on `agencies`, though a few callsites might need to do an extra db call to get the `tenants` row.

Reply:
Filed this as a separate issue, #357. Will not be addressing in this PR.